### PR TITLE
detect jar hell before installing a plugin

### LINKED ISSUE #11946: PluginManager should not install a plugin, if it would create jar hell
We now fail on ES startup if we detect jar hell (which we should keep), but I think we can probably factor out the logic to a method that takes URL[]. pluginmanager could then pass classpath of current plugins, and add the "candidate" classpath from the plugin to-be-installed.

Ideally we do all of this before touching the filesystem, so the plugin installation just fails but the overall ES installation stays working.

I will try to make a prototype.

### PR DESCRIPTION
**Pull request by rmuir**
Currently ES fails on startup if jar hell is created. But its best to fail earlier, when the user tries to install the plugin, and keep the installation healthy.

Fixes #11946

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 1]
```diff
@@ -260,6 +301,33 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
+            ZipEntry entry;
+            byte[] buffer = new byte[8192];
+            while ((entry = zipInput.getNextEntry()) != null) {
+                Path targetFile = tmp.resolve(entry.getName());
+
+                // be on the safe side: do not rely on that directories are always extracted
+                // before their children (although this makes sense, but is it guaranteed?)
+                Files.createDirectories(targetFile.getParent());
+                if (!entry.isDirectory()) {
+                    OutputStream out = Files.newOutputStream(targetFile);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: can this be a `try / with` to close the output stream?

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 2]
```diff
@@ -260,6 +301,33 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
+        try (ZipInputStream zipInput = new ZipInputStream(Files.newInputStream(zip))) {
+            ZipEntry entry;
+            byte[] buffer = new byte[8192];
+            while ((entry = zipInput.getNextEntry()) != null) {
+                Path targetFile = tmp.resolve(entry.getName());
+
+                // be on the safe side: do not rely on that directories are always extracted
+                // before their children (although this makes sense, but is it guaranteed?)
+                Files.createDirectories(targetFile.getParent());
+                if (!entry.isDirectory()) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: can this be `== false`?

**Comment by s1monw** [comment]
left some minor comments LGTM otherwise

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java (hunk 1 above)
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: so much for using @uschindler unzip code :) 

Thanks! I will take care of these

**Pull request merged**